src/instructlab/configuration.py

```diff
@@ -412,6 +547,63 @@ class Config(BaseModel):
     # model configuration
     model_config = ConfigDict(extra="ignore")
 
+    @model_validator(mode="before")
+    def fill_defaults(
+        cls, values: typing.Dict[str, typing.Any]
+    ) -> typing.Dict[str, typing.Any]:
+        defaults = {
+            "chat": {"model": DEFAULTS.DEFAULT_MODEL},
+            "serve": {
+                "model_path": DEFAULTS.DEFAULT_MODEL,
+                "llama_cpp": {"gpu_layers": -1, "max_ctx_size": 4096, "llm_family": ""},
+                "vllm": {"llm_family": "", "vllm_args": []},
+            },
+            "generate": {
+                "model": DEFAULTS.DEFAULT_MODEL,
+                "taxonomy_path": DEFAULTS.TAXONOMY_DIR,
+                "taxonomy_base": DEFAULTS.TAXONOMY_BASE,
+            },
+            "evaluate": {
+                "model": DEFAULTS.DEFAULT_MODEL,
+                "taxonomy_path": DEFAULTS.TAXONOMY_DIR,
+                "taxonomy_base": DEFAULTS.TAXONOMY_BASE,
+            },
+            "train": {
+                "model_path": DEFAULTS.MODEL_REPO,
+                "data_path": DEFAULTS.DATASETS_DIR,
+                "ckpt_output_dir": DEFAULTS.CHECKPOINTS_DIR,
+                "data_output_dir": DEFAULTS.INTERNAL_DIR,
+                "max_seq_len": 4096,
+                "max_batch_len": 10000,
+                "num_epochs": 10,
+                "effective_batch_size": 3840,
+                "save_samples": 250000,
+                "lora_quantize_dtype": None,
+                "lora_rank": 4,
+                "nproc_per_node": 1,
+                "deepspeed_cpu_offload_optimizer": False,
+                "additional_args": {},
+                "is_padding_free": False,
+            },
+            "version": CONFIG_VERSION,
+        }
+
+        for key, value in defaults.items():
+            if values.get(key) is None:
+                values[key] = value
+
+        return values
+
+
+def finish_cfg_section(
+    defaults: dict, values: typing.Dict[str, typing.Any]
+) -> typing.Dict[str, typing.Any]:
+    for key, value in defaults.items():
+        if values.get(key) is None:
```

Comment:
"values" vs. "value" collision is a little tricky to read. could be `set_values` instead

Reply:
Wondering if maybe instead of validating that it's `null`, we should validate that it's present. A user may want to override a default non-`null` value with a `null` - is it never the case?